odoo/addons/base/models/ir_model.py

```diff
@@ -1867,66 +1859,76 @@ def _reflect_constraint(self, model, conname, type, definition, module, message=
         """ Reflect the given constraint, and return its corresponding record
             if a record is created or modified; returns ``None`` otherwise.
             The reflection makes it possible to remove a constraint when its
-            corresponding module is uninstalled. ``type`` is either 'f' or 'u'
+            corresponding module is uninstalled. ``type`` is either 'f', 'i', or 'u'
             depending on the constraint being a foreign key or not.
         """
         if not module:
             # no need to save constraints for custom models as they're not part
             # of any module
             return
-        assert type in ('f', 'u')
-        cr = self._cr
-        query = """ SELECT c.id, type, definition, message->>'en_US' as message
-                    FROM ir_model_constraint c, ir_module_module m
-                    WHERE c.module=m.id AND c.name=%s AND m.name=%s """
-        cr.execute(query, (conname, module))
-        cons = cr.dictfetchone()
-        if not cons:
-            query = """ INSERT INTO ir_model_constraint
-                            (name, create_date, write_date, create_uid, write_uid, module, model, type, definition, message)
-                        VALUES (%s,
-                                now() AT TIME ZONE 'UTC',
-                                now() AT TIME ZONE 'UTC',
-                                %s, %s,
-                                (SELECT id FROM ir_module_module WHERE name=%s),
-                                (SELECT id FROM ir_model WHERE model=%s),
-                                %s, %s, %s)
-                        RETURNING id"""
-            cr.execute(query, (conname, self.env.uid, self.env.uid, module, model._name, type, definition, Json({'en_US': message})))
-            return self.browse(cr.fetchone()[0])
+        assert type in ('f', 'u', 'i')
+        rows = self.env.execute_query_dict(SQL(
+            """SELECT c.id, type, definition, message->'en_US' as message
+            FROM ir_model_constraint c, ir_module_module m
+            WHERE c.module = m.id AND c.name = %s AND m.name = %s
+            """, conname, module
+        ))
+        if not rows:
+            [[cons_id]] = self.env.execute_query(SQL(
+                """
+                INSERT INTO ir_model_constraint
+                    (name, create_date, write_date, create_uid, write_uid, module, model, type, definition, message)
+                VALUES (%s,
+                        now() AT TIME ZONE 'UTC',
+                        now() AT TIME ZONE 'UTC',
+                        %s, %s,
+                        (SELECT id FROM ir_module_module WHERE name=%s),
+                        (SELECT id FROM ir_model WHERE model=%s),
+                        %s, %s, %s)
+                RETURNING id
+                """, conname, self.env.uid, self.env.uid, module, model._name, type, definition, Json({'en_US': message})
+            ))
+            return self.browse(cons_id)
```

Comment:
Should you use the ORM here (Same for the UPDATE) or it is not ready yet?

Reply:
Reflection stuff is done during initialization of the registry: not ready yet.